server/service/mdm.go

```diff
@@ -1116,6 +1117,179 @@ func isAppleProfileID(profileIDOrUUID string) (uint, bool) {
+		return &newMDMConfigProfileResponse{Err: err}, nil
+	}
+	defer ff.Close()
+
+	fileExt := filepath.Ext(req.Profile.Filename)
+	if isApple := strings.EqualFold(fileExt, ".mobileconfig"); isApple {
+		cp, err := svc.NewMDMAppleConfigProfile(ctx, req.TeamID, ff)
+		if err != nil {
+			return &newMDMConfigProfileResponse{Err: err}, nil
+		}
+		return &newMDMConfigProfileResponse{
+			ProfileID: fmt.Sprint(cp.ProfileID),
+		}, nil
+	}
+
+	if isWindows := strings.EqualFold(fileExt, ".xml"); isWindows {
+		profileName := strings.TrimSuffix(filepath.Base(req.Profile.Filename), fileExt)
+		cp, err := svc.NewMDMWindowsConfigProfile(ctx, req.TeamID, profileName, ff)
+		if err != nil {
+			return &newMDMConfigProfileResponse{Err: err}, nil
+		}
+		return &newMDMConfigProfileResponse{
+			ProfileID: cp.ProfileUUID,
+		}, nil
+	}
+
+	err = svc.NewMDMUnsupportedConfigProfile(ctx, req.TeamID, req.Profile.Filename)
+	return &newMDMConfigProfileResponse{Err: err}, nil
+}
+
+func (svc *Service) NewMDMUnsupportedConfigProfile(ctx context.Context, teamID uint, filename string) error {
+	if err := svc.authz.Authorize(ctx, &fleet.MDMConfigProfileAuthz{TeamID: &teamID}, fleet.ActionWrite); err != nil {
+		return ctxerr.Wrap(ctx, err)
+	}
+
+	// this is required because we need authorize to return the error, and
+	// svc.authz is only available on the concrete Service struct, not on the
+	// Service interface so it cannot be done in the endpoint itself.
+	return &fleet.BadRequestError{Message: "Couldn't upload. The file should be a .mobileconfig or .xml file."}
+}
+
+func (svc *Service) NewMDMWindowsConfigProfile(ctx context.Context, teamID uint, profileName string, r io.Reader) (*fleet.MDMWindowsConfigProfile, error) {
+	if err := svc.authz.Authorize(ctx, &fleet.MDMConfigProfileAuthz{TeamID: &teamID}, fleet.ActionWrite); err != nil {
+		return nil, ctxerr.Wrap(ctx, err)
+	}
+
+	// check that Windows MDM is enabled - the middleware of that endpoint checks
+	// only that any MDM is enabled, maybe it's just macOS
+	if err := svc.VerifyMDMWindowsConfigured(ctx); err != nil {
+		err := fleet.NewInvalidArgumentError("profile", fleet.WindowsMDMNotConfiguredMessage).WithStatus(http.StatusBadRequest)
+		return nil, ctxerr.Wrap(ctx, err, "check windows MDM enabled")
+	}
+
+	var teamName string
+	if teamID > 0 {
+		tm, err := svc.EnterpriseOverrides.TeamByIDOrName(ctx, &teamID, nil)
+		if err != nil {
+			return nil, ctxerr.Wrap(ctx, err)
+		}
+		teamName = tm.Name
+	}
+
+	b, err := io.ReadAll(r)
+	if err != nil {
+		return nil, ctxerr.Wrap(ctx, &fleet.BadRequestError{
+			Message:     "failed to read Windows config profile",
+			InternalErr: err,
+		})
+	}
+
+	cp := fleet.MDMWindowsConfigProfile{
+		TeamID: &teamID,
+		Name:   profileName,
+		SyncML: b,
+	}
+	if err := cp.ValidateUserProvided(); err != nil {
+		// this is not great, but since the validations are shared between the CLI
+		// and the API, we must make some changes to error message here.
+		msg := err.Error()
```

Comment:
to be addressed later, but I still think that function I implemented is not great, should we be switching on the error type instead here?